pkg/datapath/loader/cache.go

```diff
@@ -72,6 +109,113 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
 	o.baseHash = newHash
 }
 
+// serialize finds the channel that serializes builds against the same hash.
+// Returns the channel and whether or not the caller needs to compile the
+// datapath for this hash.
+func (o *objectCache) serialize(hash string) (fq *serializer.FunctionQueue, found bool) {
+	o.Lock()
+	defer o.Unlock()
+
+	fq, compiled := o.compileQueue[hash]
+	if !compiled {
+		fq = serializer.NewFunctionQueue(1)
+		o.compileQueue[hash] = fq
+	}
+	return fq, compiled
+}
+
+func (o *objectCache) lookup(hash string) (string, bool) {
+	o.Lock()
+	defer o.Unlock()
+	path, exists := o.toPath[hash]
+	return path, exists
+}
+
+func (o *objectCache) insert(hash, objectPath string) {
+	o.Lock()
+	defer o.Unlock()
+	o.toPath[hash] = objectPath
+}
+
+// build attempts to compile and cache a datapath template object file
+// corresponding to the specified endpoint configuration.
+func (o *objectCache) build(ctx context.Context, cfg *templateCfg, hash string) error {
+	templatePath := filepath.Join(o.workingDirectory, defaults.TemplatesDir, hash)
+	headerPath := filepath.Join(templatePath, common.CHeaderFileName)
+	objectPath := filepath.Join(templatePath, endpointObj)
+
+	if err := os.MkdirAll(templatePath, defaults.StateDirRights); err != nil {
+		return err
```

Comment:
Please embed the operation performed and value for templatePath into the error as debugging the error will be very difficult otherwise